File: src/lib/planImport.ts
import { invoke } from "@tauri-apps/api/core";

import type { PlanStatus } from "./plans";

export type PlanImportCandidate = {
  slug: string;
  title: string;
  external: string;
  engine: string;
  derivedStatus: PlanStatus;
  completed: number;
  total: number;
  warning?: string;
};

export type PlanImportResult = {
  slug: string;
  planPath: string;
  status: PlanStatus;
  skipped: boolean;
  warning?: string;
};

/** Detect importable external plans (unlinked openspec/changes/ folders). */
export async function planImportDetect(projectPath: string): Promise<PlanImportCandidate[]> {
  return invoke<PlanImportCandidate[]>("plan_import_detect", { projectPath });
}

/** Import confirmed candidates by writing .basebuild/plans/<slug>/plan.md records. */
export async function planImportApply(
  projectPath: string,
  slugs: string[],
): Promise<PlanImportResult[]> {
  return invoke<PlanImportResult[]>("plan_import_apply", { projectPath, slugs });
}

File: src-tauri/src/commands/plan_import.rs
use crate::models::plan_import::{PlanImportCandidate, PlanImportResult};
use crate::services::plan_import_service;

/// Detect importable external plans in the project's `openspec/changes/`
/// directory. Returns candidates not already linked to a `.basebuild` plan
/// record. Does not modify anything on disk.
#[tauri::command]
pub fn plan_import_detect(project_path: String) -> Result<Vec<PlanImportCandidate>, String> {
    Ok(plan_import_service::detect_candidates(&project_path))
}

/// Import confirmed candidates by writing `.basebuild/plans/<slug>/plan.md`
/// records. Only candidates from the current detection pass are imported;
/// already-linked sources are skipped (idempotent). Each candidate is
/// imported independently — a failure on one does not abort the rest.
#[tauri::command]
pub fn plan_import_apply(
    project_path: String,
    slugs: Vec<String>,
) -> Result<Vec<PlanImportResult>, String> {
    Ok(plan_import_service::import_candidates(&project_path, &slugs))
}
